lib/utilities/find-addon-by-name.js

```diff
@@ -1,11 +1,34 @@
 'use strict';
 
-const find = require('ember-cli-lodash-subset').find;
+function unscope(name) {
+  if (name[0] !== '@') {
+    return name;
+  }
+
+  return name.slice(name.indexOf('/') + 1);
+}
+
+/*
+  Finds an addon given a specific name. Due to older versions of ember-cli
+  not properly supporting scoped packages it was (at one point) common practice
+  to have `package.json` include a scoped name but `index.js` having an unscoped
+  name.
 
+  Changes to the blueprints and addon model (in ~ 3.4+) have made it much clearer
+  that both `package.json` and `index.js` `name`'s should match. At some point
+  this will be "forced" and no longer optional.
+
+  This function is attempting to prioritize matching across all of the combinations
+  (in this priority order):
+
+  - package.json name matches requested name exactly
+  - index.js name matches requested name exactly
+  - unscoped (leaf portion) index.js name matches unscoped requested name
+*/
 module.exports = function findAddonByName(addons, name) {
-  function matchAddon(name, addon) {
-    return name === addon.name || (addon.pkg && name === addon.pkg.name);
-  }
+  let exactMatchFromPkg = addons.find(addon => addon.pkg && addon.pkg.name === name);
+  let exactMatchFromIndex = addons.find(addon => addon.name === name);
+  let unscopedMatchFromIndex = addons.find(addon => addon.name && unscope(addon.name) === unscope(name));
 
-  return find(addons, addon => matchAddon(name, addon));
+  return [exactMatchFromPkg, exactMatchFromIndex, unscopedMatchFromIndex].find(addon => !!addon);
```

Comment:
we should likely consider warning if there is ambiguity (even if we choose to put that warning behind a flag)